### PR TITLE
Feature/re former

### PR DESCRIPTION
**Pull request by imhta**
We just practice different ways of building the form of a rail according to the odins project.

### CODE REVIEW

File: app/views/users/new.html.erb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+
+ <%= form_for(@user, as: :user, url: users_path, authenticity_token: form_authenticity_token)do |user|%>
+  <%= render 'shared/error_messages' %>
+  <%= user.label(:username, 'Username') %>
+  <%= user.text_field(:username)%>
```
Comment (mosaaleb) [inline app/views/users/new.html.erb]: When I tried to render the new user page `/users/new` there was error `First argument in form cannot contain nil or be empty` meaning that the `@user` is nill.

HINT: you haven't created a new user in the `new` action to be available from the view file.

File: app/views/users/new.html.erb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+
+ <%= form_for(@user, as: :user, url: users_path, authenticity_token: form_authenticity_token)do |user|%>
+  <%= render 'shared/error_messages' %>
+  <%= user.label(:username, 'Username') %>
```
Comment (mosaaleb) [inline app/views/users/new.html.erb]: We don't need `authenticity_token`, `as` nor `url` options for form_for.

Check [this](https://api.rubyonrails.org/classes/ActionView/Helpers/FormHelper.html#method-i-form_for) in the arguments section.

File: app/views/users/edit.html.erb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+ <%= form_for(@user, as: :user, url: user_path, authenticity_token: form_authenticity_token)do |user|%>
+   <%= render 'shared/error_messages' %>
+  <%= user.label(:username, 'Username') %>
```
Comment (mosaaleb) [inline app/views/users/edit.html.erb]: The same also applies here we don't need authenticity token or as or url arguments.

File: app/views/users/edit.html.erb [hunk 4]
```diff
@@ -0,0 +1,10 @@
...
+  <%= user.text_field(:username)%>
+  <%= user.label(:email, 'Email address' ) %>
+   <%= user.text_field(:email)%>
+  <%= user.label(:password, 'Password' ) %>
+   <%= user.password_field(:password)%>
+  <%= user.submit( 'Submit' ) %>
+<% end %>
```
Comment (mosaaleb) [inline app/views/users/edit.html.erb]: By default rails doens't populate `password_field` value, that's why the password value won't show on the edit form. We can show it by using `<%= user.password_field(:password, value: @user.password)%>`

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    end
+
+    def show
+    end
+
+    private
+    def whitelisted_post_params
+        params.require(:user).permit(:username,:email,:password)
+    end
+end
```
Comment (mosaaleb) [inline app/controllers/users_controller.rb]: For better naming it's better to change `whitelisted_post_params` to `whitelisted_user_params`

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    end
+
+    def edit
+        @user=User.find(params[:id])
+    end
+
+    def update
+        @user=User.find(params[:id])
+        @user.update(whitelisted_post_params)
+
```
Comment (mosaaleb) [inline app/controllers/users_controller.rb]: for `update` action:
- [ ] `redirect_to edit_user_path` if user was saved
- [ ] `render :edit` if user was not saved

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -0,0 +1,40 @@
+# Users Controller
+class UsersController < ApplicationController
+  def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(whitelisted_user_params)
+    #  @user=User.new(username: params[:username], email: params[:email], password: params[:password])
```
Comment (zenott) [inline app/controllers/users_controller.rb]: Please remove this commented out code.

**Pull request merged**